### ISSUE TITLE
Slow exported model when running on TPU

### ISSUE THREAD

**Issue by rodrigonogueira4** [google-research/text-to-text-transfer-transformer#138]
I followed the instructions on T5's Colab to export a T5-base model and serve it on a TPU. However, I noticed that inference very slow: ~26 secs for a batch of 128 examples, each with 512 input tokens. Any idea what is going on?

I'm sharing my Colab in case you want to take a look:
https://colab.research.google.com/drive/1F31tPk4Cdq9cL5M6rVcbnSGNeG6s26Qv

**Comment by adarob** [comment]
That does sound a high for base, but can you try a smaller batch size?

**Comment by rodrigonogueira4** [comment]
Thanks for your answer, here are some results:
| batch_size |examples/sec|
| ---------------|------------------|
| 1 | 1.6 |
| 32 | 4.3  |
| 128 | 4.9 |

It suspect it is using CPU. At training time, I get ~128 examples/sec on a TPU v2-8 so we should expect ~256 examples/sec on inference, right?

I also set `disable_tpu=False` when exporting the model but it had no effect: https://github.com/google-research/text-to-text-transfer-transformer/blob/35a9fc06951ee2c4dc7ed0d5b1623f96377a682b/t5/models/mtf_model.py#L363

I confirm that it is using a `TPUEstimator` instance when exporting the model.
However, one weird message during export is "eval_on_tpu ignored because use_tpu is False".

**Comment by rodrigonogueira4** [comment]
It looks like that part of the answer is here: https://www.tensorflow.org/api_docs/python/tf/compat/v1/estimator/tpu/TPUEstimator

> export_saved_model exports 2 metagraphs, one with saved_model.SERVING, and another with saved_model.SERVING and saved_model.TPU tags. At serving time, these tags are used to select the appropriate metagraph to load.

> Before running the graph on TPU, the TPU system needs to be initialized. If TensorFlow Serving model-server is used, this is done automatically. If not, please use session.run(tpu.initialize_system()).

However, if I include the "tpu" tag when loading the model, it complains it coudn't find any:
```
import tensorflow.compat.v1 as tf
tf.reset_default_graph()
sess = tf.Session(TPU_ADDRESS)
sess.run(tf.tpu.initialize_system())
meta_graph_def = tf.saved_model.loader.load(sess, ["serve", "tpu"], saved_model_path)
signature_def = meta_graph_def.signature_def["serving_default"]
```

```
RuntimeError: MetaGraphDef associated with tags 'serve', 'tpu' could not be found in SavedModel. To inspect available tag-sets in the SavedModel, please use the SavedModel CLI: `saved_model_cli`
available_tags: [{'serve'}]
```

**Comment by adarob** [comment]
Can you report the speeds if you don't pass TPU_ADDRESS to the session?

**Comment by adarob** [comment]
If you set disable_tpu to false, it should export both metagraphs but you'll need to specify the TPU one. I didn't see improved performance and it caused other problems but please try that ad well.

**Comment by rodrigonogueira4** [comment]
> Can you report the speeds if you don't pass TPU_ADDRESS to the session?

Without TPU_ADDRESS, batch_size=32, and seq_len=512, it gives 0.33 examples/sec. When passing TPU_ADDRESS, we get 4.3 examples/sec. So it is seems that it is using the TPU, but I don't know why it so much slower than training and eval. Tokenization should not be bottleneck, right?

**Comment by rodrigonogueira4** [comment]
> If you set disable_tpu to false, it should export both metagraphs but you'll need to specify the TPU one. I didn't see improved performance and it caused other problems but please try that ad well.

I tried, but when I do:
```
meta_graph_def = tf.saved_model.loader.load(session, ["serve", "tpu"], args.model_dir)
```
It throws:
```
KeyError: "The name 'mtf_local_init_op' refers to an Operation not in the graph."
```

The `saved_model_cli` confirms that the TPU graph was exported, though:
```
saved_model_cli show --dir gs://neuralresearcher_data/covid/data/model_exp304/export/1584971193
The given SavedModel contains the following tag-sets:
serve
serve, tpu
```

**Comment by adarob** [comment]
Typically, a slowdown is expected due to the fact that it needs to do autoregressive sampling during inference whereas it can compute all tokens in parallel during training. However, I noticed your output size is 1 token, which makes this a bit more surprising. Nevertheless, it could still be due to some sampling overhead.

**Comment by rodrigonogueira4** [comment]
I tried with greedy decoding and got similar times.

Also, I tried different devices and below are the speeds with batch_size=16:
|Device | examples/sec
| ---------|---------|
|CPU | 0.3|
| GPU P4 | 13|
| GPU V100 | 22|
|TPU v2-8| 4|
|TPU v3-8| 8|

Note: these speeds are the best of 3 trials.
It seems that there is a noticeable overhead for TPUs.

Thanks for your help debugging this. I will keep you posted if I find where the bottleneck is.

**Comment by jhseu** [comment]
From reading your codelab, it should still be using CPU. You get pretty good examples/sec because the CPU attached to the TPU has a lot of cores. To use TPU you'll need to load the SavedModel with the tags `["serve", "tpu"]`. That graph ensures that it's placed on TPU as well as that you're utilizing all the cores.

The error that you're getting:
`KeyError: "The name 'mtf_local_init_op' refers to an Operation not in the graph."`

Is from the operation created in Mesh TensorFlow: 
https://github.com/tensorflow/mesh/blob/76ad1300d7ac0de8a8ef0bca321e7c80cbb20abd/mesh_tensorflow/transformer/utils.py#L457

I don't know enough about the software stack to know why it's being excluded when saving... my read of the code is that it should be there.

**Comment by adarob** [comment]
Thanks @jhseu that makes a lot of sense! If someone could debug the TPU serving, that would be wonderful, but unfortunately I won't have time to work on it for now.

**Comment by adarob** [comment]
I have updated the Colab in #146 to recommend using GPU until someone adds/fixes our TPU support.

**Comment by rodrigo-eai** [comment]
Great, thanks!

**Comment by rodrigonogueira4** [comment]
Great, thanks!

**Comment by jinoobaek-qz** [comment]
Is making serving work on TPU something someone is looking at? If not, do you have some pointers into how I should try and troubleshoot?

**Comment by adarob** [comment]
I don't know if anyone actively working on it. You will need to look at the `export_model` fn  in mesh_tensoflow: https://github.com/tensorflow/mesh/blob/fb41c91333e6cfd20f1e27b8f0d197f460f2fddc/mesh_tensorflow/transformer/utils.py#L1507

One thing you'll need to do is set this to True: https://github.com/tensorflow/mesh/blob/fb41c91333e6cfd20f1e27b8f0d197f460f2fddc/mesh_tensorflow/transformer/utils.py#L1256

However I'm honestly not sure what else needs to be done to make it actually work.

**Comment by adarob** [comment]
@toponado may be able to help!

**Comment by jinoobaek-qz** [comment]
Thank you for the pointers.

Do you know of any working example of mesh-tensorflow based transformer or any model serving?

**Comment by adarob** [comment]
Not with TPU. Also, just fyi, there are some people now (as of today) working on this internally (e.g., @karishmamalkan).

**Comment by jinoobaek-qz** [comment]
@adarob @karishmamalkan checking in. Is there a ball-park estimate of when we might get a solution for TPU inference using SavedModel?

**Comment by adarob** [comment]
Sorry, we are still trying to get support from the TF team to make this possible. @nshazeer may have a better idea of when it could happen.

**Comment by jinoobaek-qz** [comment]
Thanks for opening the issue and replying.

**Comment by adarob** [comment]
Closing in favor of #335.
